### PR TITLE
[mlir][Affine] Let affine.[de]linearize_index omit outer bounds

### PR DESCRIPTION
**Pull request by krzysz00**
The affine.delinearize_index and affine.linearize_index operations, as currently defined, require providing a length N basis to [de]linearize N values. The first value in this basis is never used during lowering and is unused during lowering. (Note that, even though it isn't used during lowering it can still be used to, for example, remove length-1 outputs from a delinearize).

This dead value makes sense in the original context of these operations, which is linearizing or de-linearizing indexes to memref<>s, vector<>s, and other shaped types, where that outer bound is avaliable and may be useful for analysis.

However, other usecases exist where the outer bound is not known. For example:

    %thread_id_x = gpu.thread_id x : index
    %0:3 = affine.delinearize_index %thread_id_x into (4, 16) : index,index, index

In this code, we don't know the upper bound of the thread ID, but we do want to construct the ?x4x16 grid of delinearized values in order to further partition the GPU threads.

In order to support such usecases, we broaden the definition of affine.delinearize_index and affine.linearize_index to make the outer bound optional.

In the case of affine.delinearize_index, where the number of results is a function of the size of the passed-in basis, we augment all existing builders with a `hasOuterBound` argument, which, for backwards compatibilty and to preserve the natural usage of the op, defaults to `true`. If this flag is true, the op returns one result per basis element, if it is false, it returns one extra result in position 0.

We also update existing canonicalization patterns (and move one of them into the folder) to handle these cases. Note that disagreements about the outer bound now no longer prevent delinearize/linearize cancelations.

### CODE REVIEW

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 1]
```diff
@@ -4503,62 +4504,81 @@ LogicalResult AffineVectorStoreOp::verify() {
...
+
   return success();
 }
 
 LogicalResult
 AffineDelinearizeIndexOp::fold(FoldAdaptor adaptor,
                                SmallVectorImpl<OpFoldResult> &result) {
+  // If we won't be doing any division or modulo (no basis or the one basis
+  // element is purely advisory), simply return the input value.
+  if (getStaticBasis().size() == static_cast<size_t>(hasOuterBound())) {
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4577]: I am not sure I follow this condition?

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 2]
```diff
@@ -4736,6 +4741,14 @@ LogicalResult AffineLinearizeIndexOp::verify() {
 }
 
 OpFoldResult AffineLinearizeIndexOp::fold(FoldAdaptor adaptor) {
+  // No indices linearizes to zero.
+  if (getMultiIndex().empty())
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4745]: I think we should disallow this case in the verifier.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I think the folding of the patterns with single basis is making the logic a bit more hard to follow. Left a few comments from this round of review.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp (hunk 2 above)
Comment (krzysz00) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4745]: I think we should allow this case and define it to be 0. It follows naturally from the general idea of linearization.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 3]
```diff
@@ -4576,6 +4600,20 @@ AffineDelinearizeIndexOp::fold(FoldAdaptor adaptor,
   return success();
 }
 
+SmallVector<OpFoldResult> AffineDelinearizeIndexOp::getEffectiveBasis() {
+  OpBuilder builder(getContext());
+  if (hasOuterBound()) {
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4605]: I am trying to understand why this is not just
```
auto mixedBasis = getMixedBasis(...);
if (!hasOuterBound()) {
  return llvm::to_vector(mixedBasis.drop_front());
}
return mixedBasis;
```

**Review by MaheshRavishankar — APPROVED** [review]
As far as I can read through this, this is mostly just regarding book-keeping of whether the outer bound exists or not. I still think it would be good to just drop the outer bound and call it a day.

Looks good to me though.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp (hunk 3 above)
Comment (krzysz00) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4605]: Because I was feeling clever and didn't want to pay for a reallocation I didn't need

**Comment by krzysz00** [comment]
The outer bound is important for backwards compatibility and can provide useful information in some cases
